File: NinePSharp.Messages.FSharp/messages/Rlcreate.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Rlcreate =
    { Size: uint32
      Tag: uint16
      Qid: Qid
      Iounit: uint32 }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Rlcreate
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Rlcreate span &offset
            writeQid this.Qid span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Iounit)

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let qid = readQid data &offset
        let iounit = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        { Size = size; Tag = tag; Qid = qid; Iounit = iounit }

File: NinePSharp/messages/Rlcreate.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Rlcreate : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Rlcreate;
    public ushort Tag { get; }
    public Qid Qid { get; }
    public uint Iounit { get; }

    public Rlcreate(uint size, ushort tag, Qid qid, uint iounit)
    {
        Size = size;
        Tag = tag;
        Qid = qid;
        Iounit = iounit;
    }

    public Rlcreate(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        Qid = data.ReadQid(ref offset);
        Iounit = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Rlcreate);
        int offset = NinePConstants.HeaderSize;
        span.WriteQid(Qid, ref offset);
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Iounit);
    }
}
